src/rust/inetstack/protocols/layer4/tcp/established/sender.rs

```diff
@@ -121,35 +130,161 @@ impl Sender {
+                        self.send_next_seq_no.modify(|s| s + SeqNumber::from(1));
+
+                        // Add the probe byte (as a new separate buffer) to our unacknowledged queue.
+                        let unacked_segment = UnackedSegment {
+                            bytes: Some(buf.clone()),
+                            initial_tx: Some(cb.get_now()),
+                        };
+                        self.unacked_queue.push(unacked_segment);
+
+                        // Note that we loop here *forever*, exponentially backing off.
+                        // TODO: Use the correct PERSIST mode timer here.
+                        let mut timeout: Duration = Duration::from_secs(1);
+                        loop {
+                            // Create packet.
+                            let mut header: TcpHeader = cb.tcp_header();
+                            header.seq_num = self.send_next_seq_no.get();
+                            cb.emit(header, Some(buf.clone()));
+
+                            match win_sz_watched.wait_for_change(Some(timeout)).await {
+                                Ok(_) => break,
+                                Err(Fail { errno, cause: _ }) if errno == libc::ETIMEDOUT => timeout *= 2,
+                                Err(_) => unreachable!(
+                                    "either the ack deadline changed or the deadline passed, no other errors are possible!"
+                                ),
+                            }
+                        }
+                    } else {
+                        // TODO: Nagle's algorithm - We need to coalese small buffers together to send MSS sized packets.
+                        // TODO: Silly window syndrome - See RFC 1122's discussion of the SWS avoidance algorithm.
+
+                        // We have some window, try to send some or all of the segment.
+                        // If the segment started out as 0 bytes of length, we still need to send it as a FIN.
+                        let sent_bytes: usize = self.send_segment(&mut segment, &mut cb);
+                        // If we sent at least a byte and the segment is now empty, then we are done.
+                        if sent_bytes > 0 && segment.len() == 0 {
+                            break;
+                        }
+                        // Otherwise, wait until something limiting the window changes and then try again to finish sending
+                        // the segment.
+                        futures::select_biased! {
+                            _ = send_unacked_watched.wait_for_change(None).fuse() => (),
+                            _ = self.send_next_seq_no.wait_for_change(None).fuse() => (),
+                            _ = win_sz_watched.wait_for_change(None).fuse() => (),
+                            _ = cwnd_watched.wait_for_change(None).fuse() => (),
+                            _ = ltci_watched.wait_for_change(None).fuse() => (),
+                        };
```

Comment:
Is this the priority order (left to right) in which you intend to process these futures? If yes, then this is good, else we could use `select!` instead.

Reply:
yes, I think priority order is important.